docs/rfc/000-External-plugin.md

```diff
@@ -0,0 +1,202 @@
+    }
+  }
+}
+
+type KnownPayloadTemplates = 'Card_1'
+interface MetadataDetail {
+  // points to the JSON schema to validate if it is valid
+  schema?: URL
+  preview?: SupportedPayloadPreviews
+}
+interface SupportedPayloadPreviews extends Record<KnownPayloadTemplates, URL> {
+  prefer?: KnownPayloadTemplates
+}
+```
+
+Here is an example:
+
+```jsonc
+{
+  "manifest_version": 1,
+  // this will look for "name" in the current i18n language
+  // If you don't want to do i18n, "name": "My plugin name" is also OK
+  "name": "__locales:name",
+  "description": "__locales:description",
+  "publisher": "__locales:publisher",
+  // Sign this manifest file so people will trust me (if my GPG key is famous)!
+  // If you want to sign other resources, add it to the integrity list below
+  // DISCUSS: I think it's better to have users to sign by their Mask account, aka ECC key pair.
+  "code_sign": "./sign.gpg",
+  "integrity": {
+    // I can provide hash to enforce security
+    // Opt-in feature.
+    "./preview/Card_1.json": "sha384-oqVuAfXRKap7fdgcCY5uykM6+R9GqQ8K/uxy9rx7HNQlGYl1kPzQho1wx4JwY8wC"
+    // Now with sign.gpg and this hash, we can make sure ./preview/Card_1.json is trustable
+  },
+  "i18n": {
+    "en": "./en.json",
+    "zh": "./zh.json",
+    "ja": "./ja.json"
+  },
+  "metadata": {
+    // In Mask it will be plugin:example.com/my-plugin:kind:1
+    "kind1:1": {
+      "schema": "./kind1-v1.schema.json",
+      "preview": {
+        "Card_1": "./preview/Card_1.json"
+      }
+    }
+  },
+  "contribution": {
+    // This allows to add a new badge in
+    // the composition dialog once user added the plugin
+    "composition": {
+      "icon": "./badge.svg",
+      "target": "./compose.html"
+    }
+  }
+}
+```
+
+### i18n
+
+i18n files should follow the i18next format.
+
+### Permissions
+
+You might notice that there is no permission in the manifest because all permissions are designed to be requested dynamically.
+
+### The "payload_preview" part
+
+To avoid code execution in the Mask Network extension itself and still render a plugin UI,
+we can provide some common templates allowing developers to interpolate with.
+
+Here is an example of `./preview/Card_1.json`
+
+```jsonc
+{
+  "plugin": {
+    // $ refers to the payload
+    "title": "$.title",
+    "description": {
+      // So if the __locale:preview_description is
+      // "There are still {{ balance }} in the {{ desc }}
+      // It will call the contract, get the result and plug the value in
+      "key": "__locale:preview_description",
+      "map": {
+        "desc": "$.description",
+        "balance": "#.balance"
+      }
+    },
+    "action_text": "__locale:open_to_check",
+    "action_url": "./popup.html"
+  },
+  "vars": {
+    "balance": {
+      "type": "contract",
+      "abi": "./my-contract.abi",
+      "method": "queryDetail",
+      "args": ["someVal", 0, "$.id"]
+    }
+  }
+}
+```
+
+### API in popups
+
+#### Mask.setMetadata(key: string, value: object)
```

Comment:
i want's need add an `namespace` parameter in the method

```suggestion
#### Mask.setMetadata(namespace: string, key: string, value: object)
```

identify different plugin.

Reply:
why you need a symbol here?

Reply:
internal use?

Reply:
this is part of our public API and it is going to send to other pages so you cannot transfer a symbol